### PR TITLE
[IOTDB-4420] Monitor DataNode's remain disk space through cluster heartbeat

### PR DESCRIPTION
**Pull request by CRZbulabula**
Updates:

1. Add disk_full_threshold parameter at which a DataNode will be set to ReadOnly status if the remaining disk space is lower.
2. All DataNodes report their current status periodically, so that the ConfigNode-leader always knows the correct status for each DataNode even if the leadership changed.
3. Update user doc

See [DataNode State Machine](https://apache-iotdb.feishu.cn/docx/doxcnALFW1Oxg7wIBkaDWNE6nfh) doc for more details.

### CODE REVIEW

File: node-commons/src/main/java/org/apache/iotdb/commons/conf/CommonConfig.java [hunk 1]
```diff
@@ -279,23 +304,26 @@ public NodeStatus getNodeStatus() {
...
-          new RuntimeException("System mode is set to READ_ONLY"));
-    } else {
-      logger.info("Set system mode from {} to {}.", status, newStatus);
-    }
+    logger.info("Set system mode from {} to {}.", status, newStatus);
     this.status = newStatus;
+
+    switch (newStatus) {
+      case Full:
+        logger.warn("Change system status to Full! The disk is running out!");
```
Comment (SpriCoder) [inline node-commons/src/main/java/org/apache/iotdb/commons/conf/CommonConfig.java:318]: Why not break here?

**Review by SpriCoder — CHANGES_REQUESTED** [review]

File: node-commons/src/main/java/org/apache/iotdb/commons/conf/CommonConfig.java (hunk 1 above)
Comment (CRZbulabula) [reply node-commons/src/main/java/org/apache/iotdb/commons/conf/CommonConfig.java:318]: fixed~

**Review by SpriCoder — APPROVED** [review]
Only one minor suggestion.

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/NodeManager.java [hunk 2]
```diff
@@ -670,17 +678,23 @@ public boolean isNodeRemoving(int dataNodeId) {
...
   }
 
-  public void setNodeRemovingStatus(int dataNodeId, boolean isRemoving) {
+  public void setNodeRemovingStatus(TDataNodeLocation dataNodeLocation) {
     DataNodeHeartbeatCache cache =
-        (DataNodeHeartbeatCache) configManager.getNodeManager().getNodeCacheMap().get(dataNodeId);
+        (DataNodeHeartbeatCache)
+            configManager.getNodeManager().getNodeCacheMap().get(dataNodeLocation.getDataNodeId());
     if (cache != null) {
-      cache.setRemoving(isRemoving);
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/NodeManager.java:682]: No, DataNode will report itself as Removing during heartbeat

File: node-commons/src/main/java/org/apache/iotdb/commons/conf/CommonConfig.java [hunk 3]
```diff
@@ -279,23 +304,27 @@ public NodeStatus getNodeStatus() {
...
+        break;
+      case ReadOnly:
+        logger.error(
+            "Change system status to ReadOnly! Only query statements are permitted!",
+            new RuntimeException("System mode is set to READ_ONLY"));
+        break;
+      case Removing:
+        logger.info(
+            "Change system status to Removing! The current Node is being removed from cluster!");
+    }
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply node-commons/src/main/java/org/apache/iotdb/commons/conf/CommonConfig.java:316]: Fixed

File: node-commons/src/main/java/org/apache/iotdb/commons/cluster/NodeStatus.java [hunk 4]
```diff
@@ -29,6 +29,9 @@ public enum NodeStatus {
   /** Node is in removing */
   Removing("Removing"),
 
+  /** Node disk is running out */
+  Full("Full"),
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply node-commons/src/main/java/org/apache/iotdb/commons/cluster/NodeStatus.java]: The Full status is removed

File: thrift/src/main/thrift/datanode.thrift [hunk 5]
```diff
@@ -199,7 +199,8 @@ struct THeartbeatResp {
   2: required string status
   3: optional map<common.TConsensusGroupId, bool> judgedLeaders
   4: optional i16 cpu
-  5: optional i16 memory
+  5: optional double memory
+  6: optional double disk
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply thrift/src/main/thrift/datanode.thrift]: Fixed

File: confignode/src/assembly/resources/conf/iotdb-confignode.properties [hunk 6]
```diff
@@ -342,4 +342,20 @@ target_config_nodes=127.0.0.1:22277
...
+
+####################
+### Disk Monitor
+####################
+
+
+# Threshold at which DataNode is set to Full status
+# i.e. DataNode will no longer be allocated more partitions when its status is set to full
+# Datatype: double(percentage)
+# full_threshold=5.0
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply confignode/src/assembly/resources/conf/iotdb-confignode.properties]: Fixed

File: confignode/src/assembly/resources/conf/iotdb-confignode.properties [hunk 7]
```diff
@@ -342,4 +342,20 @@ target_config_nodes=127.0.0.1:22277
...
\ No newline at end of file
+# ratis_log_appender_buffer_size_max = 4194304
+
+
+####################
+### Disk Monitor
+####################
+
+
+# Threshold at which DataNode is set to Full status
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply confignode/src/assembly/resources/conf/iotdb-confignode.properties]: Fixed

File: confignode/src/assembly/resources/conf/iotdb-confignode.properties [hunk 8]
```diff
@@ -342,4 +342,20 @@ target_config_nodes=127.0.0.1:22277
...
+# i.e. DataNode will no longer be allocated more partitions when its status is set to full
+# Datatype: double(percentage)
+# full_threshold=5.0
+
+# Threshold at which DataNode is set to ReadOnly status
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply confignode/src/assembly/resources/conf/iotdb-confignode.properties]: Fixed

File: confignode/src/assembly/resources/conf/iotdb-confignode.properties [hunk 9]
```diff
@@ -342,4 +342,20 @@ target_config_nodes=127.0.0.1:22277
...
+# i.e. DataNode will no longer be able to perform write requests
+# Datatype: double(percentage)
+# read_only_threshold=1.0
```
[the remark this answers is not in the document]
Comment (CRZbulabula) [reply confignode/src/assembly/resources/conf/iotdb-confignode.properties]: Fixed

File: thrift/src/main/thrift/datanode.thrift [hunk 10]
```diff
@@ -197,9 +197,18 @@ struct THeartbeatReq {
...
+  3: optional string statusReason
+  4: optional map<common.TConsensusGroupId, bool> judgedLeaders
+  5: optional TLoadSample loadSample
+}
+
+struct TLoadSample {
+  // Percentage of occupied cpu in DataNode
+  1: required i16 cpuOccupancyRatio
+  // Percentage of occupied memory space in DataNode
+  2: required double memoryOccupancyRatio
```
Comment (Beyyes) [inline thrift/src/main/thrift/datanode.thrift:209]: How about change `Occupancy` to `Usage`

File: docs/UserGuide/Maintenance-Tools/Maintenance-Command.md [hunk 11]
```diff
@@ -190,20 +190,13 @@ Total line number = 2
...
-|     4|Unknown|127.0.0.1|6669|            0|              0|
-|     5|Running|127.0.0.1|6671|            0|              0|
-+------+-------+---------+----+-------------+---------------+
-Total line number = 3
-It costs 0.009s
-```
+- **Running**: The DataNode is running properly and can be read and written
+- **Unknown**: The DataNode doesn't report heartbeat properly, the ConfigNode considers the DataNode as unreadable and un-writable
+- **Removing**: The DataNode is being removed from the cluster and cannot be read or written
+- **ReadOnly**: The remaining disk space of DataNode is lower than disk_full_threshold(default is 1%), the DataNode can't write or synchronize data anymore
```
Comment (Beyyes) [inline docs/UserGuide/Maintenance-Tools/Maintenance-Command.md:199]:
```suggestion
- **ReadOnly**: The remaining disk space of DataNode is lower than disk_full_threshold(default is 5%), the DataNode can't write or synchronize data anymore
```

File: docs/zh/UserGuide/Maintenance-Tools/Maintenance-Command.md [hunk 12]
```diff
@@ -188,24 +188,17 @@ Total line number = 2
...
-|     4|Unknown|127.0.0.1|6669|            0|              0|
-|     5|Running|127.0.0.1|6671|            0|              0|
-+------+-------+---------+----+-------------+---------------+
-Total line number = 3
-It costs 0.009s
-```
+- **Running**: DataNode 正常运行，可读可写
+- **Unknown**: DataNode 未正常上报心跳，ConfigNode 将认为该 DataNode 不可读写
+- **Removing**: DataNode 正在移出集群，不可读写
+- **ReadOnly**: DataNode 磁盘剩余空间低于 disk_full_threshold（默认 1%），此 DataNode 将不再能写入，不再能同步数据
```
Comment (Beyyes) [inline docs/zh/UserGuide/Maintenance-Tools/Maintenance-Command.md:197]:
```suggestion
- **ReadOnly**: DataNode 磁盘剩余空间低于 disk_full_threshold（默认 5%），此 DataNode 将不再能写入，不再能同步数据
```

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/node/BaseNodeCache.java [hunk 13]
```diff
@@ -34,11 +34,10 @@ public abstract class BaseNodeCache {
   /** SlidingWindow stores the heartbeat sample data */
   final LinkedList<NodeHeartbeatSample> slidingWindow = new LinkedList<>();
 
-  /** Node status, using for `show cluster` command */
+  /** The current status of the Node */
   volatile NodeStatus status = NodeStatus.Unknown;
-
-  /** Represent that if this node is in removing status */
-  volatile boolean removing;
+  /** The reason why lead to the current NodeStatus(for showing cluster) */
```
Comment (Beyyes) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/node/BaseNodeCache.java:39]: add a blank

**Review by Beyyes — COMMENTED** [review]
LGTM

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/node/BaseNodeCache.java (hunk 13 above)
Comment (CRZbulabula) [reply confignode/src/main/java/org/apache/iotdb/confignode/manager/node/BaseNodeCache.java]: Fixed

File: docs/zh/UserGuide/Maintenance-Tools/Maintenance-Command.md (hunk 12 above)
Comment (CRZbulabula) [reply docs/zh/UserGuide/Maintenance-Tools/Maintenance-Command.md]: Fixed

File: docs/UserGuide/Maintenance-Tools/Maintenance-Command.md (hunk 11 above)
Comment (CRZbulabula) [reply docs/UserGuide/Maintenance-Tools/Maintenance-Command.md]: Fixed

File: thrift/src/main/thrift/datanode.thrift (hunk 10 above)
Comment (CRZbulabula) [reply thrift/src/main/thrift/datanode.thrift]: Fixed

**Review by Beyyes — APPROVED** [review]

File: node-commons/src/main/java/org/apache/iotdb/commons/conf/CommonDescriptor.java [hunk 14]
```diff
@@ -104,5 +105,15 @@ public void loadCommonProps(Properties properties) {
         HandleSystemErrorStrategy.valueOf(
             properties.getProperty(
                 "handle_system_error", String.valueOf(config.getHandleSystemErrorStrategy()))));
+
+    config.setDiskSpaceWarningThreshold(
+        Double.parseDouble(
+            properties.getProperty(
+                "disk_space_warning_threshold",
```
Comment (Beyyes) [inline node-commons/src/main/java/org/apache/iotdb/commons/conf/CommonDescriptor.java:112]: maybe we can set `disk_space_warning_threshold` as a static final variable

File: confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java [hunk 15]
```diff
@@ -685,6 +687,10 @@ private TSStatus checkConfigNodeGlobalConfig(TConfigNodeRegisterReq req) {
     if (!req.getReadConsistencyLevel().equals(conf.getReadConsistencyLevel())) {
       return errorStatus.setMessage(errorPrefix + "read_consistency_level" + errorSuffix);
     }
+    if (req.getDiskSpaceWarningThreshold()
+        != CommonDescriptor.getInstance().getConfig().getDiskSpaceWarningThreshold()) {
+      return errorStatus.setMessage(errorPrefix + "disk_full_threshold" + errorSuffix);
```
Comment (Beyyes) [inline confignode/src/main/java/org/apache/iotdb/confignode/manager/ConfigManager.java:692]: change `disk_full_threshold`

**Pull request merged**